Agents/AverageSquareMetrePriceAgent/avgsqmpriceagent/kg_operations/tsclient.py

```diff
@@ -0,0 +1,72 @@
+################################################
+# Authors: Markus Hofmeister (mh807@cam.ac.uk) #    
+# Date: 23 Oct 2022                            #
+################################################
+
+# The purpose of this module is to provide functionality to use
+# the TimeSeriesClient from the JPS_BASE_LIB
+
+import agentlogging
+from avgsqmpriceagent.errorhandling.exceptions import TSException
+from avgsqmpriceagent.kg_operations.javagateway import jpsBaseLibGW
+from avgsqmpriceagent.datamodel.data import TIMECLASS
+from avgsqmpriceagent.utils.stack_configs import DB_URL, DB_USER, DB_PASSWORD
+
+# Initialise logger instance (ensure consistent logger level with `entrypoint.py`)
+logger = agentlogging.get_logger('prod')
+
+
+class TSClient:
+
+    # Create ONE JVM module view on class level and import all required java classes
+    jpsBaseLibView = jpsBaseLibGW.createModuleView()
+    jpsBaseLibGW.importPackages(jpsBaseLibView, "uk.ac.cam.cares.jps.base.query.*")
+    jpsBaseLibGW.importPackages(jpsBaseLibView, "uk.ac.cam.cares.jps.base.timeseries.*")
+
+    def __init__(self, kg_client, timeclass=TIMECLASS, rdb_url=DB_URL, 
+                 rdb_user=DB_USER, rdb_password=DB_PASSWORD):
+        """
+        Initialise TimeSeriesClient (default properties taken from environment variables)
+
+        Arguments:
+            kg_client (KGClient): KGClient object (as per `kgclient.py` (i.e. PySparqlClient))
+            timeclass: Java time class objects supported by PostgreSQL
+                       (see: https://www.jooq.org/javadoc/dev/org.jooq/org/jooq/impl/SQLDataType.html)
+            rdb_url (str): URL of relational database
+            rdb_user (str): Username for relational database
+            rdb_password (str): Password for relational database
+        """
+
+        # 1) Create an instance of a RemoteStoreClient (to retrieve RDB connection)
+        try:
+            connection = TSClient.jpsBaseLibView.RemoteRDBStoreClient(rdb_url, rdb_user, rdb_password)
+            self.conn = connection.getConnection()
```

Comment:
Instead of storing the connection object, it might be better to create a connection every time you start a query. This might be relevant #397. But I also wonder how does try-with-resources work with a python wrapper?

Reply:
Starting a new connection for every query was the default design of the TimeSeries client. Last month it got re-designed to use a connection object in order to 1) be able to use the access agent also for TimeSeries client with centrally referenced connection objects and 2) to improve performance. Currently, it seems this design becomes partially revoked; however, the methods with connection object will stay in place.

Hence, as this agent relies on recurring calls to the same Relational Database, I believe the implementation with the connection object seems suitable. Therefore, the try-with-resource should not be applicable.

Reply:
Storing a connection object can lead to some funny behaviour when running an agent in multi-thread - if one thread closes the connection then it will affect all other running queries. Given that the derivation agent allows setting `MAX_THREAD_MONITOR_ASYNC_DERIVATIONS` in config, it seems to me best to use one connection object per query, or at least per thread.

Reply:
Yes, do not store connection objects, I believe in python it will be try-catch-finally, where you do connection.close() in the finally block

Reply:
Address by [829cc40](https://github.com/cambridge-cares/TheWorldAvatar/pull/398/commits/829cc406fac7ff3f5891771a3fd12f241b62679c)

Reply:
The `@contextmanager` decorator, along with the `try`-`catch` and `yield`, seems to do what we need, so good find.